libs/json-rpc-stdio/src/lib.rs

```diff
@@ -1,35 +1,133 @@
-use jsonrpc_core::IoHandler;
-use tokio::io::{AsyncBufReadExt, AsyncRead, AsyncWrite, AsyncWriteExt};
+use jsonrpc_core::{IoHandler, MethodCall, Request};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+};
+use tokio::{
+    io::{AsyncBufReadExt, AsyncRead, AsyncWrite, AsyncWriteExt},
+    sync::{mpsc, oneshot},
+};
 
-pub async fn run(handler: &IoHandler) -> std::io::Result<()> {
-    run_with_io(handler, tokio::io::stdin(), tokio::io::stdout()).await
+#[derive(Debug)]
+pub struct Client(Arc<ClientInner>, AtomicU64);
+
+impl Client {
+    pub fn new() -> (Client, ClientAdapter) {
+        let (request_sender, request_receiver) = mpsc::channel(30);
+        let client = Client(Arc::new(ClientInner { request_sender }), Default::default());
+        let adapter = ClientAdapter { request_receiver };
+
+        (client, adapter)
+    }
+
+    pub async fn call<Req, Res>(&self, method: String, params: Req) -> jsonrpc_core::Result<Res>
+    where
+        Req: serde::Serialize,
+        Res: serde::de::DeserializeOwned,
+    {
+        let id = self.1.fetch_add(1, Ordering::Relaxed);
+        let json_params = serde_json::to_value(params).map_err(|_err| jsonrpc_core::Error::invalid_request())?;
+        let params = match json_params {
+            jsonrpc_core::Value::Array(arr) => jsonrpc_core::Params::Array(arr),
+            jsonrpc_core::Value::Object(obj) => jsonrpc_core::Params::Map(obj),
+            _ => return Err(jsonrpc_core::Error::invalid_request()),
+        };
+        let request = jsonrpc_core::MethodCall {
+            jsonrpc: Some(jsonrpc_core::Version::V2),
+            method,
+            params,
+            id: jsonrpc_core::Id::Num(id),
+        };
+        let (response_sender, response_receiver) = oneshot::channel();
+        self.0.request_sender.send((request, response_sender)).await.unwrap();
+
+        let response = response_receiver.await.unwrap();
+
+        match response {
+            jsonrpc_core::Output::Success(res) => Ok(serde_json::from_value(res.result).unwrap()),
+            jsonrpc_core::Output::Failure(res) => Err(res.error),
+        }
+    }
+}
+
+// The other side of the channels.
+pub struct ClientAdapter {
+    request_receiver: mpsc::Receiver<(MethodCall, oneshot::Sender<jsonrpc_core::Output>)>,
+}
+
+#[derive(Debug)]
+struct ClientInner {
+    request_sender: mpsc::Sender<(MethodCall, oneshot::Sender<jsonrpc_core::Output>)>,
```

Comment:
This might be better names `ClientSender`

Reply:
Yep — since we're going to remove the `Arc` that was the reason for `ClientInner`, we can even skip `ClientInner` and make request_sender a field on `Client` instead. Will do that.